core/src/main/java/io/questdb/griffin/FunctionFactoryScanner.java

```diff
@@ -0,0 +1,308 @@
+package io.questdb.griffin;
+
+import io.questdb.cairo.CairoError;
+import io.questdb.log.Log;
+import io.questdb.std.CharSequenceIntHashMap;
+import io.questdb.std.Chars;
+import io.questdb.std.str.StringSink;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.TestOnly;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.net.URL;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.lang.reflect.Modifier;
+
+// This class loads Function Factories using reflection
+// It scans the classpath for classes that implement the FunctionFactory interface
+//
+// It handles 3 different cases:
+// 1. The code is built as modules file
+// 2. The code is built as a directory
+// 3. The code is built as a single JAR file
+//
+// In case of a change all cases have to be tested
+// Loading when QuestDB runs as RT built in JDK
+// Loading when QuestDB runs as a JAR file
+// Loading when QuestDB runs as a directory, that happens when running tests from maven, IDEs
+public class FunctionFactoryScanner {
+
+    private static List<FunctionFactory> functionFactoriesCache;
+
+    @TestOnly
+    public static void clearCache() {
+        functionFactoriesCache = null;
+    }
+
+    public static Iterable<FunctionFactory> scan(String packageName, @Nullable Log log) {
+        if (functionFactoriesCache != null) {
+            return functionFactoriesCache;
+        }
+
+        try {
+            var orderMap = loadFunctionOrderMap();
+
+            // Load function factories in case the code is built as modules file
+            // This is usually the case when binaries are build with JDK baked in
+            var functionFactories = findAllClassesFromModules(packageName, log);
```

Comment:
Considering that the list may be mutated downstream, it's better to explicitly declare that we deal with `ArrayList`, not just `List`.

Reply:
It's better to declare mutating methods to accept ArrayList in this case, I can do it